### PR TITLE
Kotlin update

### PR DESCRIPTION
**Pull request by solcott**
Update Kotlin to 1.4.21 and Coroutines to 1.4.2

Also updates Gradle and some other Gradle plugins to latest.

Migrates externalEffectFlow from Buffered Channel to MutableSharedFlow

### CODE REVIEW

File: mvflow-core/src/main/kotlin/net/pedroloureiro/mvflow/MVFlow.kt [hunk 1]
```diff
@@ -197,26 +195,27 @@ private class MVFlowImpl<State, Action, Mutation, Effect>(
...
-            externalEffectChannel.send(effect)
+            externalEffectFlow.emit(effect)
         }
 
         override fun offer(effect: Effect): Boolean {
             logger.invoke("Offering external effect $effect")
-            return externalEffectChannel.offer(effect).also { accepted ->
+            return externalEffectFlow.tryEmit(effect).also { accepted ->
                 if (!accepted) {
                     logger.invoke("Channel rejected previous effect!")
```
Comment (pedroql) [inline mvflow-core/src/main/kotlin/net/pedroloureiro/mvflow/MVFlow.kt:212]: we should also update this log

File: mvflow-core/src/main/kotlin/net/pedroloureiro/mvflow/MVFlow.kt [hunk 2]
```diff
@@ -197,26 +195,27 @@ private class MVFlowImpl<State, Action, Mutation, Effect>(
...
                 }
             }
         }
     }
 
-    override fun observeEffects() = externalEffectChannel.openSubscription().consumeAsFlow()
+    override fun observeEffects() = externalEffectFlow
```
Comment (pedroql) [inline mvflow-core/src/main/kotlin/net/pedroloureiro/mvflow/MVFlow.kt:218]: Probably should use [`asSharedFlow`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/as-shared-flow.html) ?

**Review by pedroql — COMMENTED** [review]
Thank you for the PR. 

I need a bit more time to properly catch-up with shared flow docs and its nuances.

Can you tell me a few things about this PR?

1. Do you expect any behaviour changes? Is there any scenario that would be different by using a shared flow vs a channel?

2. How much experience have you got with shared flows?

3. How much have you tested/verified this work? (Something in the range of "the unit tests pass", "I tried this in my app and it seems fine", to "I have deployed this in a real app with real users and no problems arose")

Thanks again!

**Comment by pedroql** [comment]
Hi, I am applying the suggested asked here in this PR: https://github.com/pedroql/mvflow/pull/25

I'll give it a few more days for feedback and eventually merge it and close this PR.

**Comment by solcott** [comment]
In regards to the original comment.

1.  There should be no behavior changes, except that it should fix a bug I noticed in MVFlow 1.0.  

    - The bug happens when calling `observeEffects()` again after a configuration change or navigating to the next screen and coming back via the back button.  In this case I noticed that once I emitted a new effect, all previous effects were emitted again.   I fixed this by adding `debounce(100)` to the `observeEffects()` flow so I only get the latest emission.  Using a SharedFlow with replay = 0 should fix this.

2. We use SharedFlows quite a bit in our app and haven't run into any problems.
3. I ran unit tests and they passed.  I can verify in my app once I get back after New Years.

**Comment by pedroql** [comment]
Thanks, I will merge the other PR then.

Regarding

> except that it should fix a bug I noticed in MVFlow 1.0.

I'm not aware of any bug like that. I wrote a unit test that should reproduce it and it didn't work. Also, the sample in this project uses external effects to display a toast and you can change the orientation of that screen and it doesn't exhibit the behaviour you described so maybe the problem is specific with how you manage the state in your case?

Feel free to have a look here: [net/pedroloureiro/mvflow/samples/android/screens/counter](https://github.com/pedroql/mvflow/tree/master/samples/android-app/src/main/kotlin/net/pedroloureiro/mvflow/samples/android/screens/counter)

if you can reproduce the problem in that sample or still believe the problem is in the library feel free to create an issue (ideally with a sample project) and we can diagnose it.

**Pull request closed without merging**